src/transformers/models/zamba/modeling_zamba.py

```diff
@@ -0,0 +1,1758 @@
+            don't have their past key value states given to this model) of shape `(batch_size, 1)` instead of all
+            `input_ids` of shape `(batch_size, sequence_length)`.
+        inputs_embeds (`torch.FloatTensor` of shape `(batch_size, sequence_length, hidden_size)`, *optional*):
+            Optionally, instead of passing `input_ids` you can choose to directly pass an embedded representation. This
+            is useful if you want more control over how to convert `input_ids` indices into associated vectors than the
+            model's internal embedding lookup matrix.
+        use_cache (`bool`, *optional*):
+            If set to `True`, `past_key_values` key value states are returned and can be used to speed up decoding (see
+            `past_key_values`).
+        output_attentions (`bool`, *optional*):
+            Whether or not to return the attentions tensors of all attention layers. See `attentions` under returned
+            tensors for more detail.
+        output_hidden_states (`bool`, *optional*):
+            Whether or not to return the hidden states of all layers. See `hidden_states` under returned tensors for
+            more detail.
+        return_dict (`bool`, *optional*):
+            Whether or not to return a [`~utils.ModelOutput`] instead of a plain tuple.
+        cache_position (`torch.LongTensor` of shape `(sequence_length)`, *optional*):
+            Indices depicting the position of the input sequence tokens in the sequence. Contrarily to `position_ids`,
+            this tensor is not affected by padding. It is used to update the cache in the correct position and to infer
+            the complete sequence length.
+"""
+
+
+@add_start_docstrings(
+    "The bare Zamba Model outputting raw hidden-states without any specific head on top.",
+    ZAMBA_START_DOCSTRING,
+)
+class ZambaModel(ZambaPreTrainedModel):
+    """
+    Transformer decoder consisting of *config.num_hidden_layers* layers. Each layer is a [`ZambaDecoderLayer`]
+
+    Args:
+        config: ZambaConfig
+    """
+
+    def __init__(self, config: ZambaConfig):
+        super().__init__(config)
+        self.padding_idx = config.pad_token_id
+        self.vocab_size = config.vocab_size
+
+        self.embed_tokens = nn.Embedding(config.vocab_size, config.hidden_size, self.padding_idx)
+        self.block = ZambaAttentionDecoderLayer(config)
+        mamba_layers = []
+        linear_layers = []
+        self.layers_block_type = config.layers_block_type
+        for i in range(config.num_hidden_layers):
+            if config.layers_block_type[i] == "mamba":
+                mamba_layers.append(ZambaMambaDecoderLayer(config, layer_idx=i))
+            elif config.layers_block_type[i] == "attention+mamba":
+                linear_layers.append(nn.Linear(self.config.hidden_size, self.config.hidden_size, bias=False))
+                mamba_layers.append(ZambaMambaDecoderLayer(config, layer_idx=i))
```

Comment:
why is this linear layer not in the ZambaMambaDecoderLayer? Would make your life a bite easier no?

Reply:
Adding the linear layer inside `ZambaMambaDecoderLayer` would be possible but not all instantiations of `ZambaMambaDecoderLayer` are followed by a linear layer, so we opted for keeping the linear layer outside when it's present (rather than adding an attribute to `ZambaMambaDecoderLayer` that decides whether to instantiate that extra linear layer).

The linear layer is instantiated only when `ZambaMambaDecoderLayer` follows the shared transformer block`ZambaAttentionDecoderLayer` (i.e. the linear layer is between the shared transformer block and a mamba block, see fig. 2 in https://arxiv.org/pdf/2405.16712, where the linear layer is shown in red color). If `ZambaMambaDecoderLayer` follows from a previous instantiation of `ZambaMambaDecoderLayer` then there is no linear layer in between.

Reply:
sounds good. I think in general I would just recommend to have all layers, when possible in the ` self.block` , and  avoid compilcated initialization schemes (in this case, iter, next etc) in favor of simplicity! (when possible 😉 )

` Zamba` is a decoder model, but we end up with no realy `ZambaDecoderLayer` that are just called one after the other! 

I don' t have much recommendations here, let' s make sure we show the important difference of this model: the block is shared between all hybrid layers! That's important and I guess a core design of zyphra!